### PR TITLE
feat(baremetal): add offer data source

### PR DESCRIPTION
**Pull request by kindermoumoute**

### CODE REVIEW

File: scaleway/data_source_baremetal_offer_beta.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+				Description:   "Exact name of the desired offer",
+				ConflictsWith: []string{"offer_id"},
+			},
+			"offer_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "ID of the desired offer",
+				ConflictsWith: []string{"name"},
+			},
+			"allow_disabled": {
```
Comment (jerome-quere) [inline scaleway/data_source_baremetal_offer_beta.go]: I would rename to `include_disabled` but this not really clrear either

File: scaleway/data_source_baremetal_offer_beta.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Price of the offer for the next 60 minutes (a server order at 11h32 will be payed until 12h32)",
+			},
+			"price_per_month": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Price of the offer per months",
+			},
+			"quota_name": {
```
Comment (jerome-quere) [inline scaleway/data_source_baremetal_offer_beta.go]: To remove

File: scaleway/data_source_baremetal_offer_beta.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	for i := 0; i < len(res.Offers); i++ {
+		offer := res.Offers[i]
+		switch {
+		case offer.Name == d.Get("name"), offer.ID == offerID:
+			if !offer.Enable && !d.Get("allow_disabled").(bool) {
+				return fmt.Errorf("offer %s (%s) found in zone %s but is disabled. Add allow_disabled=true in your terraform config to use it.", offer.Name, offer.Name, zone)
+			}
+		default:
+			res.Offers = append(res.Offers[:i], res.Offers[i+1:]...)
+			i--
```
Comment (jerome-quere) [inline scaleway/data_source_baremetal_offer_beta.go]: I kwnow this code works but I found it hard to read 🤔

File: scaleway/data_source_baremetal_offer_beta.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	d.Set("name", offer.Name)
+	d.Set("allow_disabled", !offer.Enable)
+	d.Set("bandwidth", offer.Bandwidth)
+	d.Set("commercial_range", offer.CommercialRange)
+	d.Set("cpu", flattenCPUs(offer.CPU))
+	d.Set("disk", flattenDisks(offer.Disk))
+	d.Set("memory", flattenMemory(offer.Memory))
+	d.Set("price_per_sixty_minutes", flattenMoney(offer.PricePerSixtyMinutes))
+	d.Set("price_per_month", flattenMoney(offer.PricePerMonth))
+	d.Set("quota_name", offer.QuotaName)
```
Comment (jerome-quere) [inline scaleway/data_source_baremetal_offer_beta.go]: to remove

File: scaleway/data_source_baremetal_offer_beta.go (hunk 3 above)
Comment (kindermoumoute) [reply scaleway/data_source_baremetal_offer_beta.go]: I can, append the matching offers in a new slice, will be much more clear

File: scaleway/data_source_baremetal_offer_beta.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	}, scw.WithAllPages())
+	if err != nil {
+		return err
+	}
+
+	matches := []*baremetal.Offer(nil)
+	for _, offer := range res.Offers {
+		if offer.Name == d.Get("name") || offer.ID == offerID {
+			if !offer.Enable && !d.Get("include_disabled").(bool) {
+				return fmt.Errorf("offer %s (%s) found in zone %s but is disabled. Add allow_disabled=true in your terraform config to use it.", offer.Name, offer.Name, zone)
```
Comment (QuentinBrosse) [inline scaleway/data_source_baremetal_offer_beta.go]: `offer.Name, offer.ID, zone`?

File: scaleway/data_source_baremetal_offer_beta.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		return fmt.Errorf("%d offers found with the same name %s in zone %s", len(matches), d.Get("name"), zone)
+	}
+
+	offer := matches[0]
+	zonedID := datasourceNewZonedID(offer.ID, zone)
+	d.SetId(zonedID)
+	d.Set("offer_id", zonedID)
+	d.Set("zone", zone)
+
+	if err != nil {
```
Comment (QuentinBrosse) [inline scaleway/data_source_baremetal_offer_beta.go]: 🤔

File: scaleway/helpers_baremetal.go [hunk 7]
```diff
@@ -71,3 +71,49 @@ func getBaremetalOfferByID(baremetalAPI *baremetal.API, zone scw.Zone, offerID s
...
+		})
+	}
+	return flattenedCPUs
+}
+
+func flattenBaremetalDisks(disks []*baremetal.Disk) interface{} {
+	if disks == nil {
+		return nil
+	}
+	flattenedCPUs := []map[string]interface{}(nil)
```
Comment (QuentinBrosse) [inline scaleway/helpers_baremetal.go]: nit: `flattenedDisks`

File: scaleway/helpers_baremetal.go [hunk 8]
```diff
@@ -71,3 +71,49 @@ func getBaremetalOfferByID(baremetalAPI *baremetal.API, zone scw.Zone, offerID s
...
+		})
+	}
+	return flattenedCPUs
+}
+
+func flattenBaremetalMemory(memories []*baremetal.Memory) interface{} {
+	if memories == nil {
+		return nil
+	}
+	flattenedCPUs := []map[string]interface{}(nil)
```
Comment (QuentinBrosse) [inline scaleway/helpers_baremetal.go]: nit: `flattenedMemories`

File: scaleway/provider.go [hunk 9]
```diff
@@ -225,6 +225,7 @@ func Provider() terraform.ResourceProvider {
 			"scaleway_instance_server":         dataSourceScalewayInstanceServer(),
 			"scaleway_instance_image":          dataSourceScalewayInstanceImage(),
 			"scaleway_instance_volume":         dataSourceScalewayInstanceVolume(),
+			"scaleway_baremetal_offer_beta":    dataSourceScalewayBaremetalOfferBeta(),
```
Comment (QuentinBrosse) [inline scaleway/provider.go]: nit: move it (in the map) with other baremetal resources?

File: website/docs/d/baremetal_offer_beta.html.markdown [hunk 10]
```diff
@@ -0,0 +1,80 @@
+---
+layout: "scaleway"
+page_title: "Scaleway: scaleway_baremetal_offer_beta"
+description: |-
+  Gets information about an Baremetal offer.
```
Comment (QuentinBrosse) [inline website/docs/d/baremetal_offer_beta.html.markdown]: baremetal

File: website/docs/d/baremetal_offer_beta.html.markdown [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+In addition to all above arguments, the following attributes are exported:
+
+- `id` - The ID of the offer.
+
+- `bandwidth` - Available Bandwidth with the offer.
+
+- `commercial_range` - Commercial range of the offer.
+
+- `cpu` - A list of cpu specifications. (Structure is documented below.)
```
Comment (QuentinBrosse) [inline website/docs/d/baremetal_offer_beta.html.markdown]: IMO we need to keep the same nested lists as we have made in other resources ([example](https://www.terraform.io/docs/providers/scaleway/r/instance_server.html#root_volume)).

File: scaleway/provider.go (hunk 9 above)
Comment (kindermoumoute) [reply scaleway/provider.go]: this is the data sources map

File: website/docs/d/baremetal_offer_beta.html.markdown (hunk 11 above)
Comment (kindermoumoute) [reply website/docs/d/baremetal_offer_beta.html.markdown]: [counter example](https://www.terraform.io/docs/providers/scaleway/r/instance_security_group.html#outbound_rule)
[another counter example](https://www.terraform.io/docs/providers/google/r/container_cluster.html#type)

Looks like the general rule is:
- a list of block is documented below
- a single block is inlined

File: scaleway/provider.go (hunk 9 above)
Comment (QuentinBrosse) [reply scaleway/provider.go]: :palmface:

**Pull request merged**